Inline modules in `app/ide-desktop/`

somebody1234 opened the pull request:
### Pull Request Description
- Remove unnecessary modules
  - Remove `ts-plugin-namespace-auto-import` as it was a workaround to use the non-conventional `import *` convention
  - Remove `esbuild-plugin-copy-directories` as it is unuse
- Inline modules that are only ever used once
  - Inline `project-manager-shim` into `gui2` - it is only used during `gui2`'s dev mode
  - Inline `content-config` into `client`
- Flatten `app/ide-desktop/lib/` to `app/ide-desktop/`
- Flatten `app/ide-desktop/lib/dashboard/` to `app/dashboard/`

### Important Notes
- As mentioned above, all remaining modules have been moved up from `app/ide-desktop/lib/` to `app/ide-desktop/`. It's not ideal but I'd rather hold off on moving them anywhere else before we have a consensus on what should go where.
  - (That is to say, this may not be the final directory structure - but I figure it's fine to get *something* done so that hopefully the rest of the restructuring is simpler.)

### Code Review Notes
- There are a lot of files changed, but the vast majority are renames without changes and hence have no need to be reviewed.
- Similarly, there are a lot of lines changed, but for the most part that is unavoidable due to the flattening significantly changing the order of entries in the `package.json`.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] ~~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] ~~Unit tests have been written where possible.~~

somebody1234 committed:
Inline `assets` module

somebody1234 committed:
Remove `esbuild-plugin-copy-directories`

somebody1234 committed:
Inline `project-manager-shim` module

somebody1234 committed:
Inline `content-config` module

somebody1234 committed:
Remove `ts-plugin-namespace-auto-import`

somebody1234 committed:
Remove `types/types.d.ts`

somebody1234 committed:
Prettier

somebody1234 committed:
Fix import order

somebody1234 committed:
Remove `modules.d.ts` and `globals.d.ts` from `types` module

somebody1234 committed:
Fix build

somebody1234 committed:
Remove `ENSO_BUILD_GUI` from Electron build

somebody1234 committed:
Inline `SaveAccessTokenPayload`

somebody1234 committed:
Fix `.gitignore`

somebody1234 committed:
Flatten directory structure

somebody1234 committed:
`npm dedupe`

somebody1234 committed:
Fix paths

somebody1234 committed:
Fix build

somebody1234 committed:
Merge branch 'develop' into wip/sb/inline-modules

somebody1234 committed:
Fix ESLint

somebody1234 committed:
Suppress Vite errors

somebody1234 committed:
Fix lint error

somebody1234 committed:
Fix GUI2 E2E tests

farmaazon commented:
> Flatten app/ide-desktop/lib/ to app/

Well, when the directory structure was designed, I imagined `app` folder representing not _the application_ but rather _all applications_:
* engine (compiler and maybe project manager)
* GUI (a package common for cloud browser and electron package)
* ide-desktop - electron package using GUI above

Thence came gui and ide-desktop. Much later on, dashboard landed inside `ide-desktop` which started breaking the above schema (it should land in GUI instead... but GUI was then in rust). Also, engine never made it to this `app` folder, because none wanted to spend time on sbt reconfiguring.

But still, I think we should ultimately keep separation between GUI and electron, just for cloud-in-browser sake.

~~Also, **these changes may greatly conflict with @Frizi efforts of migrating to pnpm - a part of moving to Bazel**, mostly done as far as I'm aware. I would wait for @Frizi changes to land first, to avoid nasty conflicts. Also, all significant directory restructuring should be consulted with "Bazel" team (@hubertp and @Frizi)~~

Edit: ok, I think I'm not aware, or misremembered some arrangements, so please ignore the second part.

farmaazon commented on app/gui2/projectManagement.ts:
```diff
@@ -0,0 +1,399 @@
+ * - if the project is already in the Project Manager's location, we just open it;
+ * - if the project is in a different location, we copy it to the Project Manager's location
+ * and open it.
+ * - if the project is a bundle, we extract it to the Project Manager's location and open it. */
```
Judging by this documentation, I don't know why it is in gui2 directory. gui2 is essentially `Project View` and the project view is not responsible for importing or managing projects through PM.

Please, add some TODO that this is required only for dev build (which should be soon reworked anyway).

farmaazon commented on app/gui2/projectManagerShimMiddleware.ts:
```diff
@@ -0,0 +1,453 @@
+/** @file A simple HTTP server which serves application data to the Electron web-view. */
```
This docs seem to be plainly wrong.

farmaazon reviewed (commented):
For now, I would keep at least electron-specific packages still in `ide-desktop` (removing unnecessary `lib`)

somebody1234 committed:
Move Electron packages back into `ide-desktop`

somebody1234 committed:
Address CR comments

somebody1234 committed:
Fix `package.json`

somebody1234 committed:
Fix build

somebody1234 committed:
Fix TS project references

somebody1234 committed:
Merge branch 'develop' into wip/sb/inline-modules

somebody1234 committed:
Fix lint errors

somebody1234 committed:
Merge branch 'develop' into wip/sb/inline-modules

somebody1234 committed:
Fix build errors

somebody1234 committed:
Merge branch 'develop' into wip/sb/inline-modules

somebody1234 committed:
Minor changes to TSConfigs

somebody1234 committed:
Remove dependency on dashboard enum in `toasts.ts`

It's not worth the minimal advantages at the cost of breaking module
boundaries.

somebody1234 committed:
Fix lint errors

somebody1234 commented:
umm...
https://github.com/enso-org/enso/actions/runs/9675920069/job/26694401507#step:7:639

anyone know how to fix project references in TS?

<blockquote><img src="https://opengraph.githubassets.com/b5e0e7f01247f42def9b3812dd63fa8736472099e2e96304a262f8948cee14f5/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Inline modules in `app/ide-desktop/` · enso-org/enso@0c52059</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

somebody1234 commented:
Pinging @Frizi for visibility

Frizi commented on app/ide-desktop/client/src/preload.ts:
```diff
@@ -5,6 +5,8 @@
 
 import * as electron from 'electron'
 
+import * as dashboard from 'enso-dashboard'
```
should this be a type-only import?

Frizi commented on app/dashboard/src/components/AriaComponents/Form/components/useForm.ts:
```diff
@@ -54,7 +54,8 @@ export function useForm<
       ...options,
-      resolver: zodResolver.zodResolver(schema, { async: true }),
+      // eslint-disable-next-line no-restricted-syntax
+      resolver: zodResolver.zodResolver(schema as never, { async: true }),
```
Why is this change necessary?

somebody1234 replied:
no clue :skull: - i guess @MrFlashAccount will know more about this

MrFlashAccount replied:
@somebody1234 but this change doesn't exist on develop... not sure I can explain it either...

somebody1234 replied:
yeah i have no idea XD it's probably to do with some accidental version bump, i guess i'll take a look on monday

Frizi commented on app/dashboard/vite.config.ts:
```diff
@@ -21,10 +21,14 @@ await appConfig.readEnvironmentFromFile()
   server: { port: SERVER_PORT, headers: Object.fromEntries(common.COOP_COEP_CORP_HEADERS) },
   plugins: [
+    // @ts-expect-error It seems like there is a type error caused by differences in type
+    // definitions between Vite versions.
```
This version mismatch is now fixed in develop, those expect-errors will probably be flagged after merge.

Frizi commented on app/gui2/projectManagement.ts:
Lets move all `project-manager-shim`-related files to a subdiriectory.

Frizi commented on build/build/src/ide/web.rs:
```diff
@@ -123,7 +123,7 @@ pub fn unpacked_dir(output_path: impl AsRef<Path>, os: OS, arch: Arch) -> PathBu
 /// Computes the SHA-256 checksum of a file and writes it to a file.
 ///
-/// This is a Rust equivalent of the `app/ide-desktop/lib/client/tasks/computeHashes.mjs`.
+/// This is a Rust equivalent of the `app/client/tasks/computeHashes.mjs`.
```
There is no `app/client`, should there be `ide-desktop/client`?

Frizi commented on app/eslint.config.mjs:
I'm not a fan of this eslint config leaking into other apps now, but I guess it filters the files to dashboard anyway. We will have to tidy it up (i.e. remove most of the coutnerproductive rules) in a separate PR.

Frizi reviewed (approved):
Spotted some minor things to check or fix, but overall that this is a very good change and will simplify a bunch of things in the future 👍

somebody1234 committed:
Merge branch 'develop' into wip/sb/inline-modules

somebody1234 committed:
Clean up npm packages

somebody1234 committed:
Move GUI2 project manager shim to subdirectory

somebody1234 committed:
Address CR

somebody1234 committed:
Fix errors

somebody1234 committed:
Forgot how eslint works

somebody1234 committed:
Update codeowners

somebody1234 committed:
Fix codeowners

somebody1234 commented:
@jdunkerley i think we need the approval of someone from libs because `dataLinkSchema.json` is being moved.

also the CODEOWNERS file has been updated and that one should probably be reviewed too

somebody1234 commented:
@farmaazon @Frizi either of you happen to know why lint CI is saying this?
https://github.com/enso-org/enso/actions/runs/9714248920/job/26812883079?pr=10305#step:7:200
```
 INFO ide_ci::program::command: npm⚠️ npm ERR! code 127
 INFO ide_ci::program::command: npm⚠️ npm ERR! path /runner/_work/enso/enso/app/gui2
 INFO ide_ci::program::command: npm⚠️ npm ERR! command failed
 INFO ide_ci::program::command: npm⚠️ npm ERR! command sh -c playwright install
 INFO ide_ci::program::command: npm⚠️ npm ERR! sh: 1: playwright: not found
```

i also got it while running locally but not sure what could possibly cause this...

<blockquote><img src="https://opengraph.githubassets.com/bfb765abe0c95c819faf75bce3efa25060a77572db6a7af01808ef8b1494f005/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Inline modules in `app/ide-desktop/` · enso-org/enso@cc9493e</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

farmaazon commented:
I cannot reproduce this locally: I only have different errors in tests.

It looks really strange: the only place where I see `playwright install` is in our postinstall, but there it should be run from local package, so playwright should be visible.

somebody1234 commented:
hmm, i think it's only reproducible if you're missing certain packages, not sure which. maybe it might be reproducible if you delete `node_modules`?

> there it should be run from local package

yeah, indeed... no idea how it could possibly fail tbh. which is why i just ignored it and moved on when i encountered it

farmaazon commented:
You may check what if you add `npx` there.

After removing node modules, I get a different error:

```
 INFO ide_ci::program::command: npmℹ️ src/entrypoint.ts(4,28): error TS6305: Output file '/home/adam-praca/dev/enso/node_modules/.cache/tsc/enso/app/dashboard/src/index.d.ts' has not been built from source file '/home/adam-praca/dev/enso/app/dashboard/src/index.tsx'.
 INFO ide_ci::program::command: npm⚠️ npm ERR! Lifecycle script `typecheck` failed with error: 
 INFO ide_ci::program::command: npm⚠️ npm ERR! Error: command failed 
 INFO ide_ci::program::command: npm⚠️ npm ERR!   in workspace: enso-gui2@0.1.0 
 INFO ide_ci::program::command: npm⚠️ npm ERR!   at location: /home/adam-praca/dev/enso/app/gui2 
```

somebody1234 committed:
Merge branch 'develop' into wip/sb/inline-modules

somebody1234 committed:
`pnpm i`

somebody1234 committed:
Prettier

somebody1234 committed:
Fix lint errors

somebody1234 committed:
Fix import error

somebody1234 committed:
Add missing packages

somebody1234 committed:
Remove incorrect comment

somebody1234 committed:
Fix vitest?

somebody1234 commented:
idk what's up with CI o_O
```
Cannot find package 'yaml' imported from /runner/_work/enso/enso/app/gui2/vitest.config.ts.timestamp-1720431019684-09bd4b9a42eb1.mjs
```

somebody1234 committed:
Merge branch 'develop' into wip/sb/inline-modules

somebody1234 committed:
pnpm i

somebody1234 committed:
Install missing module

somebody1234 committed:
Merge branch 'develop' into wip/sb/inline-modules

somebody1234 committed:
Fix lint errors

somebody1234 committed:
Merge branch 'develop' into wip/sb/inline-modules

somebody1234 committed:
Fix bad merge

somebody1234 committed:
Merge branch 'develop' into wip/sb/inline-modules

somebody1234 committed:
Fix type errors

somebody1234 committed:
Finally fix YAML issue

somebody1234 committed:
Update paths

somebody1234 committed:
Fix import path

somebody1234 committed:
Defer missing environment variables detection

somebody1234 committed:
Fix CODEOWNERS

somebody1234 committed:
Fix `icons` package?
